odoo/service/model.py

```diff
@@ -138,6 +138,8 @@ def retrying(func, env):
                 if env.cr._closed:
                     raise
                 env.cr.rollback()
+                if env.registry is not odoo.registry():
+                    env.reset()
```

Comment:
I wonder whether `env.reset()` shouldn't _always_ be done in this case (and the one below, too).
This would ensure that you always restart a request from scratch when retrying it.